server/join.go

```diff
@@ -29,36 +30,118 @@ import (
 const defaultDialTimeout = 30 * time.Second
 
 // TODO: support HTTPS
-func (cfg *Config) genClientV3Config() clientv3.Config {
+func genClientV3Config(cfg *Config) clientv3.Config {
 	endpoints := strings.Split(cfg.Join, ",")
 	return clientv3.Config{
 		Endpoints:   endpoints,
 		DialTimeout: defaultDialTimeout,
 	}
 }
 
-// prepareJoinCluster send MemberAdd command to pd cluster,
-// returns pd initial cluster configuration.
-func (cfg *Config) prepareJoinCluster() (string, error) {
+func memberAdd(client *clientv3.Client, urls []string) (*clientv3.MemberAddResponse, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), defaultDialTimeout)
 	defer cancel()
 
-	client, err := clientv3.New(cfg.genClientV3Config())
+	return client.MemberAdd(ctx, urls)
+}
+
+func memberList(client *clientv3.Client) (*clientv3.MemberListResponse, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultDialTimeout)
+	defer cancel()
+
+	return client.MemberList(ctx)
+}
+
+// prepareJoinCluster send MemberAdd command to pd cluster,
+// returns pd initial cluster configuration.
+//
+// TL;TR: The join functionality is safe. With data, join does nothing, w/o data
+//        and it is not a member of cluster, join does MemberAdd, otherwise
+//        return an error.
+//
+// Etcd automatically re-join cluster if there is a data dir, so first we check
+// if there is data dir or not. With data dir, it returns an empty string(etcd
+// will get correct configurations from data dir.)
+//
+// Without data dir, we may have following cases:
+//
+//  1. a new pd joins to an existing cluster.
+//      join does: MemberAdd, MemberList, then generate initial-cluster.
+//
+//  2. a new pd joins itself
+//      join does: nothing.
+//
+//  3. an failed pd re-joins to previous cluster.
+//      join does: return an error(etcd reports: raft log corrupted, truncated,
+//                 or lost?)
+//
+//  4. a join self pd failed and it restarted with join while other peers try
+//     to connect to it.
+//      join does: nothing. (join can not detect whether it is in a cluster or
+//                 not, however, etcd will handle it safey, if there is no data
+//                 in cluster the restarted pd will join to cluster, otherwise
+//                 pd will shutdown as soon as other peers connect to it. etcd
+//                 reports: raft log corrupted, truncated, or lost?)
+//
```

Comment:
+//  4. A PD fails to join itself and it is restarted with join while other peers try
 +//     to connect to it.
 +//     What join does: nothing. (join cannot detect whether it is in a cluster or
 +//                 not, however, etcd will handle it safely. If there is no data
 +//                 in the cluster, the restarted PD will join the cluster, otherwise,
 +//                 PD will shut down as soon as other peers connect to it. etcd
 +//                 reports: raft log corrupted, truncated, or lost?)